Feature/async cell

dacky179 opened the pull request:
Cell supports async session management.
PublishNotReadyToWorkAsync and PublishActivityCompletedAsync added.
Remove Capabilities because it hides inherited member of Resource.

dacky179 committed:
Cell supports async session management.
PublishNotReadyToWorkAsync and PublishActivityCompletedAsync was added.

dacky179 committed:
Rearrange Properties and methods.
Revert unintended deletions.
Remove Capabilities because it hides inherited member ob Resource.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
     [Description("Base type for all cells within a production system")]
     public abstract class Cell : Resource, ICell
     {
+        private readonly Dictionary<Guid, TaskCompletionSource<Session>> _sessionCompletionSources =
```
What do you think about using a `ConcurrentDicrionary` here? Since we don't know how many threads might modify this in parallel 🤔

dacky179 replied:
good idea

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+        /// CancellationTokenSource that must be canceled during <see cref="OnStop"/>.
+        /// Used to cancel async operations during resource shutdown.
+        /// </summary>
+        protected CancellationTokenSource LifeCycleTokenSource = new CancellationTokenSource();
```
could be readonly, couldn't it? 🤔

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
 
-        /// <inheritdoc />
+        /// <summary>
+        /// Callback of the control system, to start an activity in the cell.
```
Great explanation to understand why the explicit implementation is important here 👍 May I suggest to make it a bit more general, as the implementation details of the process engine are not necessarily known to the developer here:
> Callback to start an activity on the cell after a `ReadyToWork` event was raised. If not otherwise explicitly required all depending components will use the interface to refer to cells, making this the right place to interject and complete tasks started by async calls to \<see cref="PublishReadyToWorkAsync(Moryx.ControlSystem.Cells.ReadyToWork)"/\>.

dacky179 replied:
took your comment

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+        void ICell.StartActivity(ActivityStart activityStart)
+        {
+            // check if session was started async
+            if (_sessionCompletionSources.TryGetValue(activityStart.Id, out var completionSource))
```
Please combine with `TryRemove` method

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+            {
+                _sessionCompletionSources.Remove(activityStart.Id);
+                if (!completionSource.TrySetResult(activityStart))
+                    Logger.Log(LogLevel.Error, $"Cannot set result of async request. [{nameof(ActivityStart)}]");
```
Please also add the session id to the log message, to allow to connect it to other session related actions

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+            StartActivity(activityStart);
+        }
+
+        /// <summary>
```
I would leave the `/// <inheritdoc />` comment here

dacky179 replied:
member is not inherited.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
-        public ICapabilities Capabilities
+		
+        /// <summary>
+        /// Callback from the control system, that the sequence was completed.
```
Again great comment to explain why this is happening here. Suggestion for a more general phrasing
>Callback to complete a sequence on the cell after an `ActivityCompleted` event was raised. If not otherwise explicitly required all depending components will use the interface to refer to cells, making this the right place to interject and complete tasks started by async calls to \<see cref="PublishActivityCompletedAsync(Moryx.ControlSystem.Cells.ActivityCompleted)"/\> .

dacky179 replied:
took your comment

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+        void ICell.SequenceCompleted(SequenceCompleted completed)
         {
-            get
+            if (_sessionCompletionSources.TryGetValue(completed.Id, out var completionSource))
```
Please combine with `TryRemove`

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
-                return _capabilities;
+                _sessionCompletionSources.Remove(completed.Id);
+                if (!completionSource.TrySetResult(completed))
+                    Logger.Log(LogLevel.Error, $"Cannot set result of async request. [{nameof(SequenceCompleted)}]");
```
PLease add the session id for tracability

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+        /// <summary>
+        /// Callback from the control system, that the sequence was completed.
+        /// </summary>
+        /// <param name="completed"></param>
```
I would leave the `/// <inheritdoc />` comment here

dacky179 replied:
It is not longer the inherited method. It must get his own documentation.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+        protected override void OnStop()
+        {
+            LifeCycleTokenSource.Cancel();
+            LifeCycleTokenSource.Dispose();
```
Careful here, if you dispose on stop, a restart to the resource management will run into issues as the completion source is not recreated in `OnStart`. I would propose implementing the OnStart method as well and putting the object initialization there 😊

dacky179 replied:
Thomas once told me there is ne reinc lifcycle in resource management. He told me everything will be build up new... maybe @Toxantron could clarify that to me.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
         public abstract void StartActivity(ActivityStart activityStart);
 
         /// <inheritdoc />
         public virtual void ProcessAborting(IActivity affectedActivity) { }
```
Do you have any thoughts on handling activities here that relate to pending task completion sources in `_sessionCompletionSources`? 🤔

dacky179 replied:
hm, the only case i see is the milisecond between sending session result async and getting the abort. its seems to be a very very low chance to hit that. but i will try

dacky179 replied:
what do you think now?

1nf0rmagician replied:
Good point, there is not much to handle. I forgot that a running activity is not in the dictionary. So maybe it can be ignored. I opened a thread to ask @Toxantron and otherwise we could remove it and proceed without it 🙈

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+        /// </summary>
+        public async Task<Session> PublishReadyToWorkAsync(ReadyToWork readyToWork, CancellationToken cancellationToken)
+        {
+            Logger.Log(LogLevel.Trace, $"PublishReadyToWorkAsync Session {readyToWork.Id} Type {readyToWork.ReadyToWorkType}, Classification {readyToWork.AcceptedClassification}, {readyToWork.Reference}");
```
Log messages should always use named placeholders and put the variables in the args behind the message 😊
[Here ](https://stackoverflow.com/questions/52200484/why-logging-doesnt-use-string-interpolation) a discussion if you are interested in background information. There should also be a compiler warning for this line 🙈

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+            // throw exception if cancellation via token was requested
+            linkedToken.ThrowIfCancellationRequested();
+            // double check if any of the combined Tokens are already canceled.
+            cancellationToken.ThrowIfCancellationRequested();
```
Let's skip the double check as there is no benefit in it  and please add the exception to the xml comment to make it clear to users that they need to take care of that exception (for both overloads of the `PublishReadyToWorkAsync` method)😊

dacky179 replied:
sandro told me he found an issue related to the LinkedTokenSource

dacky179 replied:
sandro and I were not able to finde source of his information; changed like suggested

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
+
+            var completionSource = new TaskCompletionSource<Session>(linkedToken);
+            _sessionCompletionSources.Add(readyToWork.Id, completionSource);
+            ReadyToWork!.Invoke(this, readyToWork);
```
I don't think we can be sure here that there is a subscriber for the `ReadyToWork` or the `NotReadyToWork` further down 🤔

dacky179 replied:
This would throw an exception if a resource raises a RTW before CS is attached. Its a clear programming fail to do that.

Toxantron replied:
Downside: The exception here might break the cells state machine (which was technically broken before as it should prevent the call)

Upside: No exception means you might also end up in a life lock, since you never receive a response for the RTW u raised

dacky179 replied:
I could check for null and return cancled?

dacky179 replied:
check now :-)

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -16,34 +19,120 @@ namespace Moryx.ControlSystem.Cells
             {
-                _capabilities = value;
-                this.CapabilitiesChanged?.Invoke(this, _capabilities);
+                Logger.Log(LogLevel.Information, $"PublishReadyToWorkAsync canceled! Session {readyToWork.Id} Publish NotReadyToWork");
```
See logging comment above

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
         /// </summary>
         public void PublishNotReadyToWork(NotReadyToWork notReadyToWork)
         {
+            if (_sessionCompletionSources.TryGetValue(notReadyToWork.Id, out var completionSource))
```
Please combine as suggested above

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
         public event EventHandler<NotReadyToWork> NotReadyToWork;
 
+        /// <summary>
+        /// Publish <see cref="ActivityCompleted"/> from the resource
```
In the ready to work methods you also added the excepted session types and possible exceptions 😊

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
+        /// <param name="cancellationToken"></param>
+        public Task<Session> PublishActivityCompletedAsync(ActivityCompleted activityResult, CancellationToken cancellationToken)
+        {
+            Logger.Log(LogLevel.Trace, $"PublishActivityCompletedAsync Session {activityResult.Id}, Classification {activityResult.AcceptedClassification}, {activityResult.Reference}");
```
See logging comment

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
+            // throw exception if cancellation via token was requested
+            linkedToken.ThrowIfCancellationRequested();
+            // double check if any of the combined Tokens are already canceled.
+            cancellationToken.ThrowIfCancellationRequested();
```
see above

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
+
+            var completionSource = new TaskCompletionSource<Session>();
+            _sessionCompletionSources.Add(activityResult.Id, completionSource);
+            ActivityCompleted!.Invoke(this, activityResult);
```
I don't think we can expect this to be not null necessarily🙈

dacky179 replied:
Now it must be not null since we cannot start a session if events are not wired.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -56,16 +145,55 @@ public void PublishReadyToWork(ReadyToWork readyToWork)
+            LifeCycleTokenSource.Token.ThrowIfCancellationRequested();
+
+            var completionSource = new TaskCompletionSource<Session>();
+            _sessionCompletionSources.Add(activityResult.Id, completionSource);
```
The ConcurrentDictionary has a `TryAdd` method, still we need to decide here what happens when application code falsely tries to publish an activity completed twice. If there should be an exception this should be provided in the comment; otherwise we could use the `AddOrUpdate` method and send it to the process engine as would be done with the not-async method

dacky179 replied:
i decieded to cancle new request if there is an existing. trys to return exiting task and check type of sessions are to complicated

1nf0rmagician reviewed (commented):
Despite the number of comments looks great, looking forward to seeing your updates to the PR! 😊

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
Ahh, and please update the VERSION file to the next minor 😊

dacky179 replied:
done

dacky179 committed:
Use ConcurrentDictionary.
Improve comments.
Handle AbortActivity for pending async results.
Use TryRemove.
Check ReadyToWork-event to be wired.

dacky179 committed:
Handle already running async operations.
Increase version.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -36,46 +40,78 @@ public abstract class Cell : Resource, ICell
+            var asyncResult = _sessionCompletionSources.SingleOrDefault(pair =>
+                pair.Key is ActivityCompleted completed &&
+                completed.CompletedActivity.Id == affectedActivity.Id);
+            if (asyncResult.Key != null)
```
@Toxantron is there a scenario in which a session completed will come after a ProcessAborting was called? My understanding was that there won't be any feedback coming for an activity afterwards 🤔

dacky179 committed:
More description of methods.

dacky179 commented:
@1nf0rmagician I updated everything. Please check.

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -36,46 +40,78 @@ public abstract class Cell : Resource, ICell
+            if (_sessionCompletionSources.TryRemove(completed, out var completionSource))
             {
-                _sessionCompletionSources.Remove(completed.Id);
+                // by setting the result teh related async call (PublishReadyToWorkAsync or PublishActivityCompletedAsync) will be completed.
```
Typo, and you should have a warning that numerical placeholders in log messages are discouraged and should be replaced with named placeholders (counts for all logs), i.e.
`Logger.Log(LogLevel.Error, "Cannot set result of async request for session {id}. [{name}]", completed.Id, nameof(SequenceCompleted));`

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -109,28 +150,53 @@ public Task<Session> PublishReadyToWorkAsync(ReadyToWork readyToWork)
+                // now waiting for StartActivity() or SequenceCompleted()
+                completionSource.Task.GetAwaiter().GetResult();
+            }
+            catch (TaskCanceledException e)
```
note warnings from github, you can just remove the e in the brackets 😊

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -109,28 +150,53 @@ public Task<Session> PublishReadyToWorkAsync(ReadyToWork readyToWork)
+            try
+            {
+                // now waiting for StartActivity() or SequenceCompleted()
+                completionSource.Task.GetAwaiter().GetResult();
```
I think you want to return the task here, don't you? This call would block the thread and might get you in serious trouble if two many threads get stuck here 😬

dacky179 replied:
no, I wanted to await the completion and check for cancled to publish NotReadyToWork automaticaly.

dacky179 replied:
```suggestion
            void PublishNotReadyToWorkIfCanceled(Task<Session> task)
            {
                if (task.IsCanceled)
                {
                    Logger.Log(LogLevel.Information, "PublishReadyToWorkAsync canceled! Session {sessionId} Publish NotReadyToWork", readyToWork.Id);
                    // NotReadyToWork must be wired because we raised ReadyToWork before!
                    NotReadyToWork!.Invoke(this, readyToWork.PauseSession());
                }
            }

            var task = completionSource.Task.WaitAsync(linkedToken)
                .ContinueWith(PublishNotReadyToWorkIfCanceled, CancellationToken.None);

            return completionSource.Task;
```

dacky179 replied:
@seveneleven was not a fan of ContinueWith; and we decided to use the current solution.

1nf0rmagician replied:
I talked to @Toxantron and @seveneleven. I would suggest
```c#
return completionSource.Task.ContinueWith(t => {
    if (t.IsCanceled) {
        Logger.Log(LogLevel.Information, "PublishReadyToWorkAsync canceled! Session {0} Publish NotReadyToWork", readyToWork.Id);
        NotReadyToWork!.Invoke(this,readyToWork.PauseSession());
    }
    // Not sure but you might want to return the session or the original task here 
});
```
Otherwise you have a serious deadlock risk at your hand here and the whole point of async would be mitigated😬

dacky179 replied:
done

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -109,28 +150,53 @@ public Task<Session> PublishReadyToWorkAsync(ReadyToWork readyToWork)
-            if (completionSource.Task.IsCanceled)
+
+            // check event to be wired
+            if (ReadyToWork == null)
```
Nice error handling here, there are to many places in the framework that have errog logs that don't differentiate errors already😁

1nf0rmagician commented on src/Moryx.ControlSystem/Cells/Cell.cs:
```diff
@@ -165,45 +230,77 @@ public void PublishNotReadyToWork(NotReadyToWork notReadyToWork)
+        /// <summary>
+        /// IEqualityComparer to compare different sessions in a dictionary by their Id.
+        /// </summary>
+        private class SessionComparer : IEqualityComparer<Session>
```
I like it 👍

dacky179 committed:
Change string parameters to named.

dacky179 committed:
Typo fixed

dacky179 committed:
Fix behavior of a canceled PublishReadyToWorkAsync to not block threads.
